### DISCUSSION TITLE
Can a I define a `with` on a Type field?

### DISCUSSION THREAD

**Discussion by larskoole** [rebing/graphql-laravel#1082] [Q&A] [answered]
I'm trying to get my `UserType` to eager-load a specific relation when a specific field (`role`) is being requested.

```php
// UserType.php

public function fields(): array
{
  return [ 
    'role' => [
      'type' => GraphQL::type('Role!'),
      'description' => 'The role of the user for the current business.',
      'resolve' => function (User $user) {
        return $user->businesses->where('id', currentBusiness()->id)->first()->pivot->role;
      },
      'selectable' => false,
    ],
  ];
}
```

Currently the `businesses` relation on the user is not eager-loaded so I'm getting an N+1 warning.

What I essentially want to do is define a `with` field of some sorts on the `role` field like this:
```php
// UserType.php

public function fields(): array
{
  return [ 
    'role' => [
      'type' => GraphQL::type('Role!'),
      'description' => 'The role of the user for the current business.',
      'resolve' => function (User $user) {
        return $user->businesses->where('id', currentBusiness()->id)->first()->pivot->role;
      },
      'selectable' => false,
      'with' => [
        'businesss',
      ],
    ],
  ];
}
```

Is there a way to eager-load the businesses?

**Comment by mfn** [comment]
How do your actual `resolve` look like? Did you try the stuff from https://github.com/rebing/graphql-laravel#eager-loading-relationships ?

**Reply by larskoole** [reply]
I guess exactly the same as the docs, but the problem is that in this query the `$fields->getRelations()` has no way of knowing it also needs to load the `businesses` on the user right?

```php
public function resolve($root, array $args, $context, ResolveInfo $resolveInfo, Closure $getSelectFields): LengthAwarePaginator
{
    $fields = $getSelectFields();
    return currentBusiness()
        ->users()
        ->with($fields->getRelations())
        ->select($fields->getSelect())
        ->orderBy('id')
        ->paginate($args['limit'] ?? 10);
}
```

And for obvious reasons I don't want to hardcode it in the resolve because the `UserType` is used/referenced in multiple queries/types.

The docs don't state anything about hinting an eager-load of a relation.
The only thing you can hint is to load a column by using `always => [...]`.
Essentially I need an `always` to eager-load relations.

**Reply by mfn** [reply]
Hmm, I'm sorry I don't think I can help further.

Personally I don't use `SelectFields` at all and exclusively use custom data loader using Eloquent, which can solve those problems but also is a wholly different approach and requires much more "code to make it work", but it's much more flexible (as you can control in each dataloader what and how you want to eager load stuff, even when it gets more complex with filter, 2nd or 3rd degree relations, etc.)

**Reply by larskoole** [reply]
Ah ok no worries.

Do you have any more resources on dataloaders?
Would love to see some examples on how to make them and use it with Laravel/Eloquent.

**Reply by mfn** [reply, marked as answer]
Basically the way the underlying graphql-php library suggests it, see https://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem

The realization to start returning `new Deferred()` with a closure from a resolver is key to achive the lazy-eager-loading correctly.

A commonly used library is https://github.com/overblog/dataloader-php but I found it better in my project to use a custom implementation.
